cheat-sheet.md

```diff
@@ -0,0 +1,250 @@
+
+```
+sealed trait Boolean
+
+case object True extends Boolean
+
+case object False extends Boolean
+```
+
+### Pattern matching
+Pattern matching is a mechanism for checking a value against a pattern. Pattern matching is similar to a series of if/else statements or a switch statement in other languages.
+
+```
+def numToString(num: Int): String =
+  num match {
+    case 0 => "zero"
+    case 1 => "one"
+    case 2 => "two"
+    case _ => "many"
+  }
+```
+
+### map
+Map works by applying a function to each element in the list.
+
+```
+def map[A,B](fa: F[A])(f: A => B): F[B]
+
+def map[A,B](oa: Option[A])(f: A => B): Option[B]
+
+def map[A,B](ea: Either[A])(f: A => B): Either[B]
+
+def map[A,B](ta: Try[A])(f: A => B): Try[B]
+```
+
+###flatMap
+flatMap works by applying a function that returns a sequence for each element in the list, and flattening the results into the original list.
+
+```
+flatMap[A,B](fa: F[A])(f: A => F[B]): F[B]
+
+flatMap[A,B](oa: Option[A])(f: A => Option[B]): Option[B]
+
+flatMap[A,B](ea: Either[A])(f: A => Either[B]): Either[B]
+
+flatMap[A,B](ta: Try[A])(f: A => Try[B]): Try[B]
+```
+
+###For-Comprehensions
+A for-comprehension is syntactic sugar for map, flatMap and filter operations on collections.
+
+The general form is `for {s} yield e`
+`s` is a sequence of generators and filters
+`p <- e` is a generator
+`if f` is a filter
+
+For example:
+```
+for {
+    x <- Some(3);
+    y <- Some(4)
+} yield (x + y)
+
+// result = Some(7)
+```
+
+is equivalent to:
+```
+Some(3).flatMap(x => Some(4).map(y => x + y))
+```
+
+
+### Option Data Type
+The Option data type has two cases: it can be defined (Some), or it can be undefined (None).
+
+```
+sealed trait Option[+A]
+
+case class Some[+A](get: A) extends Option[A]
+
+case object None extends Option[Nothing]
+```
+
+Example:
+```
+def mean(xs: Seq[Double]): Option[Double] =
+    if (xs.isEmpty) None
+    else Some(xs.sum / xs.length)
+```
+
+### Either Data Type
+The Either data type has only two cases which both carry a value. Either represents values that can be one of two things. 
+When we use it to indicate success or failure, by convention the `Right` constructor is reserved for the success case and `Left` is used for failure.
+
+```
+sealed trait Either[+E, +A]
+
+case class Left[+E](value: E) extends Either[E, Nothing]
+
+case class Right[+A](value: A) extends Either[Nothing, A]
+```
+
+Example:
+```
+def mean(xs: IndexedSeq[Double]): Either[String, Double] =
+    if (xs.isEmpty) Left("mean of empty list!")
+    else Right(xs.sum / xs.length)
+```
+
+### Try Data type
+The Try function is a general-purpose function we can use to convert from an exception-based API to an Option or Either data type. 
```

Comment:
this is not a quite accurate explanation of what the Try data type is. If the idea is to show how to use it maybe just paraphrase a bit what the scala doc says:
```
The Try type represents a computation that may either result in an exception, or return a successfully computed value
```

But if the intent is to explain the `Try Expression` (i.e `try{ ...} catch { ... }`  then maybe just rephrase some stamentes